### ISSUE TITLE
Adding Graph in web interface forces edge definition collections

### ISSUE THREAD

**Issue by geekygnr** [arangodb/arangodb#1529]
Hello,

I am using the web interface to define a number of different graphs. They all use the same edge collection (member) but I want each graph to show the edges between different collections.

For example if I have a collection of sports players, hockey teams, and base ball teams. I would have one graph mapping the players to hockey teams without showing the base ball connections and another only showing the base ball.

For example the hockey graph might have
hockey/jr_aaa -> hockey/siskins -> Jon Smith
But not have any base ball information.

While the baseball graph would have
baseball/house_league -> baseball/royals -> Jon Smith
But not have any hockey information.

Anyways the problem is if the hockey graph is the first one I create, when I make the baseball graph and add members to edge definitions it forces the collections to have the hockey collection and I can't change it.

Not sure if this is a bug or I am not understanding something but it is a little annoying. I am using 2.7.0 installed through apt on Ubuntu 14.04

**Comment by dothebart** [comment]
You can use vertices in several collections; however you can't use the same edge collection for several graphs.  [the city graph sample](https://docs.arangodb.com/Graphs/index.html#the_city_graph) as its used in [the documentation of the graph functions](https://docs.arangodb.com/GeneralGraphs/Functions.html) uses several edge collections and several vertex collections for one graph.

**Comment by geekygnr** [comment]
Hello,

I am still struggling with this restriction. To be honest if I just modify the _graph collection it will create the graphs as I want where I reuse edge collections.

Could you help me understand why this restriction exists?

**Comment by dothebart** [comment]
If you want the [graph to be consistent](https://docs.arangodb.com/Aql/GraphFunctions.html#graph-consistency) you must not remove edges from outside of the graph management. If you have two graphs sharing collections, this consistency can't be guaranteed.

**Comment by geekygnr** [comment]
That seems to result in a lot of collections containing the same type of data. In your cars example there were 5 collections for 2 data types: highways and cities.

I'm sorry I think I am having a problem converting mentalities from relational databases to graphing databases. Looks like I have more reading to do.

**Comment by dothebart** [comment]
The [router example](https://docs.arangodb.com/Graphs/index.html#the_city_graph) distributes the graph over several edge and vertex collections.

The [knows graph](https://docs.arangodb.com/Graphs/index.html#the-knowsgraph) only uses one edge and vertex collection each.
The [newer example traversal graph](https://docs.arangodb.com/devel/Graphs/index.html#the-traversal-graph) also has one collection each.

Think of edges being pointers similar to C-style pointers enhanced to being C++ Objects that can contain attributes (like "label" in the last example). This pointer however is bi-directional like you would do in a double linked list.

In SQL an edge collection would probably be a pivot-table. The advantage of the Graph over flat pivot tables is, that they can dynamically traverse several steps over such pivot tables ; regardless whether they're one or several pivot tables with an unknown number of steps.

ArangoDB 2.8 is going to bring [native graph traversals](https://docs.arangodb.com/devel/Aql/GraphTraversals.html); the documentation contains a graphical explanation what the algorithm does under the hood - maybe that also can better explain it.

**Comment by mchacki** [comment]
Hi,

i think i have to clarify the concept of the graph definitions a bit more in detail:

First of all we have the concept of `edge definitions`:
An `edge definition` wraps around one edge collection and adds additional restrictions for the entries in `_from` and `_to`.
As an example let's say we have vertices `users` and and edges `isFriend`.
We can use the edge definition `isFriend`, `from: [users]`,`to: [users]` which makes sure only edges from `users` to `users` are allowed, edges from `users` to `products` will be rejected.
The classical edge collections in ArangoDB do NOT check such conditions on their own.

Now we have the concept of a `graph`.
A `graph` takes a list of such `edge definitions` and makes sure their conditions remain fulfilled.
Say now we have another definition `hasBought` `from: [users]`, `to: [products]`. We can define the graph `ecommerce` using `hasBought` and `isFriend`.

And finally i think we come to the source of your confusion.
Now we 1. want to have another graph that should share the the same `edge definition` `isFriend`.
This means these graphs actually share the same piece of data, same users, same relations between them. This is totally allowed and supported with the simple restriction that `isFriend` has identical definition in both graphs (it is identical data) and it is forbidden that the second graph says: `isFriend` is `from: [users]`to `to: [customers]`
The reason for this restriction is pretty simple, `commerce` expects that targets of these edges are `users` but now finds `customers` instead. So there is a consistency problem for this graph.

And we can have a second graph that wants to only reuse the data stored in vertices (say `users`) only. Than it can define a new `edge definition` with a different name using `from: [users]` or `to: [users]`.
Than the relations are independent but the vertices are shared.


So in short what is not possible is to create two edge definitions having the same name but different `from` and `to` values.

Also please do not modify the `_graphs` collection by hand. All sanity-checks in the graph-module expect the data stored in this collection is correct and do not revalidate it every time. So inserting potentially illegal data there might cause strange behavior which is very hard to debug.
Also the `_graphs` collection might change during different versions of ArangoDB which could cause even more strange behavior.

**Comment by kibarda** [comment]
Hey! 
.. so the first use case of an ArangoDB graph would be to define a scope where consistency is to be enforced in case of a complex delete ? is that correct ? And what were the other possible technical solutions ? 

Thanks.

**Comment by kibarda** [comment]
what would be helpful for geekygnr (and other people) would be something like graphViews, the equivalent of rdbms views but on a graph. There would be no restriction enforcement as it would be used only for data selection. It could be used  directly as select, within traversal, graph visualization. They could overlap, be nested,.. 
Just an idea.

**Comment by geekygnr** [comment]
Understanding what I do now I think I need to restructure my data.

What I am left wondering is if there is a use case for an edge collection without an edge definition? If not I would recommend that the edge definitions be set when you make a new edge collection instead of when you make a graph. How it is set up right now in the UI it looks like edge definitions are associated on a per graph basis which is where my confusion came from and why I thought there was a bug that I needed to find a work around for (which led me to _graphs).

**Comment by dothebart** [comment]
I've enriched https://docs.arangodb.com/devel/Graphs/index.html with more documentation about the topics you've mentioned. Hopefully this may work better for others then ;-) 
If you have more questions, let me know.
Closing this now.
